src/lightning_app/components/auto_scaler.py

```diff
@@ -0,0 +1,529 @@
+from fastapi.security import HTTPBasic, HTTPBasicCredentials
+from lightning_utilities.core.overrides import is_overridden
+from pydantic import BaseModel
+from starlette.status import HTTP_401_UNAUTHORIZED
+
+from lightning_app.core.flow import LightningFlow
+from lightning_app.core.work import LightningWork
+from lightning_app.utilities.app_helpers import Logger
+from lightning_app.utilities.packaging.cloud_compute import CloudCompute
+
+logger = Logger(__name__)
+
+
+def _raise_granular_exception(exception: Exception) -> None:
+    """Handle an exception from hitting the model servers."""
+    if not isinstance(exception, Exception):
+        return
+
+    if isinstance(exception, HTTPException):
+        raise exception
+
+    if isinstance(exception, aiohttp.client_exceptions.ServerDisconnectedError):
+        raise HTTPException(500, "Worker Server Disconnected") from exception
+
+    if isinstance(exception, aiohttp.client_exceptions.ClientError):
+        logging.exception(exception)
+        raise HTTPException(500, "Worker Server error") from exception
+
+    if isinstance(exception, asyncio.TimeoutError):
+        raise HTTPException(408, "Request timed out") from exception
+
+    if isinstance(exception, Exception):
+        if exception.args[0] == "Server disconnected":
+            raise HTTPException(500, "Worker Server disconnected") from exception
+
+    logging.exception(exception)
+    raise HTTPException(500, exception.args[0]) from exception
+
+
+class _SysInfo(BaseModel):
+    num_workers: int
+    servers: List[str]
+    num_requests: int
+    process_time: int
+    global_request_count: int
+
+
+class _BatchRequestModel(BaseModel):
+    inputs: List[Any]
+
+
+def _create_fastapi(title: str) -> FastAPI:
+    fastapi_app = FastAPI(title=title)
+
+    fastapi_app.add_middleware(
+        CORSMiddleware,
+        allow_origins=["*"],
+        allow_credentials=True,
+        allow_methods=["*"],
+        allow_headers=["*"],
+    )
+
+    fastapi_app.global_request_count = 0
+    fastapi_app.num_current_requests = 0
+    fastapi_app.last_process_time = 0
+
+    @fastapi_app.middleware("http")
+    async def current_request_counter(request: Request, call_next):
+        if not request.scope["path"] == "/api/predict":
+            return await call_next(request)
+        fastapi_app.global_request_count += 1
+        fastapi_app.num_current_requests += 1
+        start_time = time.time()
+        response = await call_next(request)
+        process_time = time.time() - start_time
+        fastapi_app.last_process_time = process_time
+        fastapi_app.num_current_requests -= 1
+        return response
+
+    @fastapi_app.get("/", include_in_schema=False)
+    async def docs():
+        return RedirectResponse("/docs")
+
+    @fastapi_app.get("/num-requests")
+    async def num_requests() -> int:
+        return fastapi_app.num_current_requests
+
+    return fastapi_app
+
+
+class _LoadBalancer(LightningWork):
```

Comment:
is it what @aniketmaurya and el. did with Muse App?

Reply:
It's pretty much the same as muse's, but I marked it as private since I don't want to expose it (as it's not usable as a standalone lightning work currently.).

In the future, we could consider making `_LoadBalancer` usable as a standalone component and rename it `LoadBalancer` (public).

cc @aniketmaurya @ethanwharris